datalad_next/runners/run.py

```diff
@@ -0,0 +1,64 @@
+"""
+"""
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from pathlib import Path
+from queue import Queue
+from subprocess import (
+    DEVNULL,
+    TimeoutExpired,
+)
+from typing import (
+    IO,
+)
+
+from datalad.runner.nonasyncrunner import _ResultGenerator
+
+from . import (
+    Protocol,
+    ThreadedRunner,
+)
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: Protocol,
+    *,
+    cwd: Path | None = None,
+    input: int | IO | bytes | Queue[bytes | None] | None = None,
+    # only generator protocols make sense for timeout, and timeouts are
+    # only checked when the generator polls
+    timeout: float | None = None,
+) -> dict | _ResultGenerator:
+    runner = ThreadedRunner(
+        cmd=cmd,
+        protocol_class=protocol_class,
+        stdin=DEVNULL if input is None else input,
+        cwd=cwd,
+        timeout=timeout,
+    )
+    try:
+        yield runner.run()
+    finally:
+        # if we get here the subprocess has no business running
+        # anymore. When run() exited normally, this should
```

Comment:
> \# if we get here the subprocess has no business running
        \# anymore. [...]

I am not sure that that is necessarily true. That would only be true if the code that is executed in the run-context has a way to ensure that the subprocess is actually done.  Either by means of communication with the process or by a `Popen.wait()`. If the context-code can ensure that, then it would be justified to always terminate() and possibly kill() the subprocess.

Is it the intended semantics that the code that is executed in the run-context ensures that the subprocess is actually done?!

Reply:
> I am not sure that that is necessarily true.

I think so. This is a context manager. User using a context manager can and should have the expectation that the resources managed by in a context are freed/closed/whatever on exit.

What that means for an internal implementation is another business. but a context manager is typically used to encapsulate all the complexities that you point out, such that a user need not constantly be aware of everything.